File: DAI/BasicDAIPlayer-iOS/objc/BasicDAIPlayer/VideoPropertiesViewController.m
//
//  VideoPropertiesViewController.m
//  BasicDAIPlayer
//
//  Copyright © 2023 Brightcove, Inc. All rights reserved.
//

#import "VideoPropertiesViewController.h"


@implementation VideoPropertiesViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
}

- (void)setupPlaybackController
{
    id<BCOVPlaybackController> playbackController = ({
        IMASettings *imaSettings = [IMASettings new];
        imaSettings.language = NSLocale.currentLocale.localeIdentifier;
        IMAAdsRenderingSettings *adsRenderingSettings = [IMAAdsRenderingSettings new];
        adsRenderingSettings.linkOpenerDelegate = self;
        adsRenderingSettings.linkOpenerPresentingController = self;

        BCOVDAIAdsRequestPolicy *adsRequestPolicy = [BCOVDAIAdsRequestPolicy videoPropertiesAdsRequestPolicy];

        id<BCOVPlaybackSessionProvider> daiSessionProvider = [self.manager createDAISessionProviderWithSettings:imaSettings
                                                                                           adsRenderingSettings:adsRenderingSettings
                                                                                               adsRequestPolicy:adsRequestPolicy
                                                                                                    adContainer:self.playerView.contentOverlayView
                                                                                                 viewController:self
                                                                                                 companionSlots:nil
                                                                                        upstreamSessionProvider:nil
                                                                                                        options:nil];

        id<BCOVPlaybackController> playbackController = [self.manager createPlaybackControllerWithSessionProvider:daiSessionProvider
                                                                                                     viewStrategy:nil];
        playbackController.delegate = self;
        playbackController.autoPlay = YES;
        playbackController.autoAdvance = YES;
        playbackController;
    });

    self.playerView.playbackController = playbackController;

    self.playbackController = playbackController;
}

- (BCOVVideo *)updateVideo:(BCOVVideo *)video
{
    BCOVVideo *updatedVideo = [video update:^(id<BCOVMutableVideo>  _Nonnull mutableVideo) {

        NSDictionary *adProperties = @{
            kBCOVDAIVideoPropertiesKeySourceId: kViewControllerGoogleDAISourceId,
            kBCOVDAIVideoPropertiesKeyVideoId: kViewControllerGoogleDAIVideoId
        };

        NSMutableDictionary *propertiesToUpdate = mutableVideo.properties.mutableCopy;
        [propertiesToUpdate addEntriesFromDictionary:adProperties];
        mutableVideo.properties = propertiesToUpdate;
    }];

    return updatedVideo;
}

@end

File: DAI/BasicDAIPlayer-iOS/swift/BasicDAIPlayer/VideoPropertiesViewController.swift
//
//  ViewController.swift
//  BasicDAIPlayer
//
//  Copyright © 2024 Brightcove, Inc. All rights reserved.
//

import UIKit
import GoogleInteractiveMediaAds
import BrightcoveDAI


final class VideoPropertiesViewController: BaseViewController {

    override func setupPlaybackController() {
        guard let sdkManager = BCOVPlayerSDKManager.shared(),
              let fps else { return }

        let imaSettings = IMASettings()
        imaSettings.language = NSLocale.current.languageCode!

        let adsRenderingSettings = IMAAdsRenderingSettings()
        adsRenderingSettings.linkOpenerDelegate = self
        adsRenderingSettings.linkOpenerPresentingController = self

        let adsRequestPolicy = BCOVDAIAdsRequestPolicy.videoProperties();

        // BCOVDAIPlaybackSessionDelegate defines -willCallIMAAdsLoaderRequestAdsWithRequest:
        // which allows us to modify the IMAStreamRequest object before it is used to load ads.
        let daiPlaybackSessionOptions = [ kBCOVDAIOptionDAIPlaybackSessionDelegateKey: self ]

        let daiSessionProvider = sdkManager.createDAISessionProvider(with: imaSettings,
                                                                     adsRenderingSettings: adsRenderingSettings,
                                                                     adsRequestPolicy: adsRequestPolicy,
                                                                     adContainer: playerView!.contentOverlayView,
                                                                     viewController: self,
                                                                     companionSlots: nil,
                                                                     upstreamSessionProvider: fps,
                                                                     options: daiPlaybackSessionOptions)

        guard let playerView,
              let playbackController = sdkManager.createPlaybackController(with: daiSessionProvider,
                                                                           viewStrategy: nil) else {
            return
        }

        playbackController.delegate = self
        playbackController.isAutoPlay = true
        playbackController.isAutoAdvance = true

        playerView.playbackController = playbackController

        self.playbackController = playbackController
    }

    override func updateVideo(_ video: BCOVVideo) -> BCOVVideo {
        let updatedVideo = video.update { (mutableVideo: BCOVMutableVideo?) in
            guard let mutableVideo else {
                return
            }

            if var updatedProperties = mutableVideo.properties {
                updatedProperties[kBCOVDAIVideoPropertiesKeySourceId] = kGoogleDAISourceId
                updatedProperties[kBCOVDAIVideoPropertiesKeyVideoId] = kGoogleDAIVideoId

                mutableVideo.properties = updatedProperties
            }
        }

        return updatedVideo
    }
}
